### PR TITLE
Immediately remove messages flagged as assault or harassment

### PR DESCRIPTION
**Pull request by neil-marcellini**
cc @thienlnam
### Details

Now when you flag a message as assault or harassment it will be immediately removed from view. The message won't show in the chat, LHN, or on the search page.

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/289225
PROPOSAL: N/A

### Tests

First check out the corresponding Auth PR https://github.com/Expensify/Auth/pull/8058, build Auth and restart it.

1. Sign in with any account A
2. On another device sign in with account B
3. Send a message from B to A, saying "flag my next message"
4. Send another message
5. As account A, flag the message as assault or harassment
6. Verify that the message is immediately removed from the chat and is no longer visible
7.  Verify that you see a whisper message from concierge saying `Thanks for flagging that comment for the Expensify team to review. You will no longer see that message and the Expensify team will take the necessary steps to resolve the situation.`
8. Verify that you do not see the flagged message in the LHN section for account B and see "flag my next message" instead
9. Open the search page and search for account B
10. Verify that you do not see the flagged message on the search page and see "flag my next message" instead
11. Refresh the page and verify that the flagged message is still removed and the whisper is still visible.
12. Verify the LHN and search page as before
13. As account B, verify that the flagged message is removed from the chat
14. Verify that you see a whisper message from concierge saying `One of your messages has been flagged. I’ve hidden the message as a potential violation of Rule #2 (Don’t Ruin It For Everyone Else), and it will be reviewed. I will contact you when I have completed my review.`
1. Verify that you do not see the flagged message in the LHN section for account A and see "flag my next message" instead
8. Open the search page and search for account A
9. Verify that you do not see the flagged message on the search page and see "flag my next message" instead

- [ ] Verify that no errors appear in the JS console

### Offline tests

Same as the tests above, but go offline, flag the comment, then go back online.
**Note:** This doesn't match the normal offline behavior where the pending "delete" has a strikethrough style, but I think it's best to make an exception for this case because it's not exactly a delete, and more importantly if someone says something awful you want it gone right away, even if you're offline.

### QA Steps

Same as tests.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
I'm only testing on web because these changes are platform independent.
<details>
<summary>Web</summary>

Online

https://github.com/Expensify/App/assets/26260477/b1ff2d7b-fdd0-40ee-b5f9-4f5f687477ce

Offline

https://github.com/Expensify/App/assets/26260477/52155b88-9b64-4890-b29b-3c66ee5170b5

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by neil-marcellini** [comment]
I did notice a small issue which should probably be fixed. Although I'm successfully hiding the message in the chat, it's still showing up in the LHN from the report.lastMessageText, as you can see in the offline test video. It would also show up on the search or new chat page.

I'll leave it up to you guys whether we need to fix that before merging this. I could write some hacky code here in App to use the previous action message if the latest action is pendingRemove, or I could fix it in Web-Expensify. I'm not sure what's best but I would like to do it in a follow up so we can get the majority of this feature done in time.

**Comment by abdulrahuman5196** [comment]
@neil-marcellini  I don't have access to auth PR mentioned in the tests, Do we need to wait for it to merge for testing this PR?

File: src/libs/ReportActionsUtils.js
Comment (thienlnam) [inline src/libs/ReportActionsUtils.js:328]: Let's use moderationDecision here (we're migrating to that since we'll only return the last decision anyways)
https://github.com/Expensify/Auth/blob/87fb42c3df3638181e670b8b01d29922acc09d2e/auth/command/Get.cpp#L268-L268

**Review by thienlnam — CHANGES_REQUESTED** [review]

File: src/libs/ReportActionsUtils.js
Comment (thienlnam) [reply src/libs/ReportActionsUtils.js:328]: Hmm actually, I wonder if we'd have to update FlagComment then as well

Comment (thienlnam) [reply src/libs/ReportActionsUtils.js:328]: Actually yeah, I don't think this would work offline if we made this change - I added a clean up issue here https://github.com/Expensify/Expensify/issues/292452 but feel free to ignore the change here

**Comment by thienlnam** [comment]
> @neil-marcellini I don't have access to auth PR mentioned in the tests, Do we need to wait for it to merge for testing this PR?

@abdulrahuman5196 Yeah, you won't be able to do the full tests until the auth PR is deployed (happening later today), but this should work locally/offline. Like if you flag a message as assault it should be hidden

File: src/libs/ReportActionsUtils.js
Comment (neil-marcellini) [reply src/libs/ReportActionsUtils.js:328]: Ok cool thanks!

**Comment by neil-marcellini** [comment]
I just realized we don't need to hold this on the Auth deploy. It won't break anything without the Auth changes, but it won't work until the Auth changes are deployed.

**Comment by neil-marcellini** [comment]
Ok @thienlnam @dangrous @roryabraham this is ready for review now. The Auth PR is merged but isn't on production quite yet. It should be soon and I'll try to comment here once it's deployed.

File: src/libs/OptionsListUtils.js
Comment (thienlnam) [inline src/libs/OptionsListUtils.js:92]: I'm surprised this doesn't impact any other methods here, there isn't any place that uses allReportActions???

Comment (thienlnam) [reply src/libs/OptionsListUtils.js:92]: Also how come the change to lastReportActions? It doesn't seem like it is used in any of your changes

File: src/libs/OptionsListUtils.js
Comment (thienlnam) [inline src/libs/OptionsListUtils.js:395]: Right now, we're doing this for every report. 
Since this flow is not the main flow can we update the logic so that by default, we don't look for the last non whisper unless we have to?

So basically

Check the last action had a moderation decision

If not, proceed as normal
Otherwise, find the last visible reportAction

**Review by thienlnam — CHANGES_REQUESTED** [review]

File: src/libs/OptionsListUtils.js
Comment (neil-marcellini) [reply src/libs/OptionsListUtils.js:92]: Yeah a couple things were surprising when I started looking at this, as you mentioned.

1. allReportActions is unused (idk why lint didn't catch that)
2. lastReportActions are not actually the "last" report actions for each report. They don't come pre-sorted in the object, they need to be sorted as we do on the ReportScreen with getSortedReportActions. I figured I should fix that while I'm making changes.

File: src/libs/OptionsListUtils.js
Comment (neil-marcellini) [reply src/libs/OptionsListUtils.js:395]: We need to check the last action that is not a whisper, because otherwise we're looking to see if the whisper is moderated which will always be false.

_.find will stop as soon as it finds a non whisper action, so it's only going to look through n actions where n is the number of whispers in a row from the latest action moving back in time. I think that's ok.

Comment (thienlnam) [reply src/libs/OptionsListUtils.js:395]: Hmm yeah, I guess there isn't a great way to avoid this

File: src/libs/OptionsListUtils.js
Comment (thienlnam) [inline src/libs/OptionsListUtils.js:385]: Can we just create an isWhisperReportAction in ReportActionUtils and then use it here and below to dry this up a bit?

**Review by thienlnam — CHANGES_REQUESTED** [review]
Just one last comment, but rest looks good

File: src/libs/OptionsListUtils.js
Comment (neil-marcellini) [reply src/libs/OptionsListUtils.js:385]: Sure. We can use it elsewhere too.

File: src/pages/home/report/ReportActionItem.js
Comment (thienlnam) [inline src/pages/home/report/ReportActionItem.js:418]: Why this change?

Comment (neil-marcellini) [reply src/pages/home/report/ReportActionItem.js:418]: Lodash get is unnecessary here and this is faster and just as safe. Tim taught me that.

**Review by thienlnam — APPROVED** [review]

**Comment by thienlnam** [comment]
My bad, forgot the checklist as I was on mobile but I did test this PR locally. @neil-marcellini could you please remove the label as I can't on mobile 🥲

**Comment by neil-marcellini** [comment]
I'm un-assigning @abdulrahuman5196 because we had to get this done quickly and so he wasn't able to test or review.

**Comment by abdulrahuman5196** [comment]
Thank you for notifying. Good to have priority changed merged sooner. I was actually under the impression auth PR was going to be merged yesterday and I planned to test this PR today. Apologies if any delay from my end.

**Comment by aldo-expensify** [comment]
@neil-marcellini I think this may be causing this deploy blocker: https://github.com/Expensify/App/issues/20940

**Comment by aldo-expensify** [comment]
@neil-marcellini @dangrous 

Do you think we should implement a similar filter for the last report action in Auth: https://github.com/Expensify/Auth/blob/58d8c605f8d820bf6283fb98101798e05287aa26/auth/lib/Report.cpp#L4385-L4416

Right now it doesn't ignore the report action if it has been reported.

**Comment by dangrous** [comment]
Oh yeah, I think that's correct - we should mirror the change there. Let me know if you agree @neil-marcellini

**Comment by neil-marcellini** [comment]
Yeah I think that would be better. @thienlnam discussed something like that previously but we decided to do the quick and dirty version on the front end. What problem is solved by doing the filtering on the backend?

**Comment by aldo-expensify** [comment]
> What problem is solved by doing the filtering on the backend?

I'm not 100% sure, but it feels like the backend now calculates an incorrect version of the "last message" of a report which we use for LHN, and the front end has to correct it with the code front end code added. It just sounds dirtier to me, and harder to maintain because the last message in the report coming from the backend is not reliable anymore.

**Comment by neil-marcellini** [comment]
Great explanation, I agree!

**Comment by aldo-expensify** [comment]
I think this https://github.com/Expensify/App/pull/20223#issuecomment-1595419997 may be also causing that the `report.lastActorAccountID` is wrong, see this:

https://github.com/Expensify/App/pull/20949#issuecomment-1597612599


I'll try to fix it in the backend.

File: src/libs/ReportActionsUtils.js
Comment (dangrous) [reply src/libs/ReportActionsUtils.js:330]: ah nice catch!

File: src/libs/OptionsListUtils.js
Comment (aldo-expensify) [inline src/libs/OptionsListUtils.js:90]: We never clean up `allSortedReportActions` so it keeps gathering garbage between sessions. This confused me a lot since I was able to find reportActions that were not in Onyx. This is also a memory leak.

@neil-marcellini 🙏 fix

**Pull request merged**
